Starting by setting up the environment.

In [5]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

import keras
from keras import layers # Layers api for sequential model

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
print(os.listdir("../input/heart-disease-uci"))

# Any results you write to the current directory are saved as output.

['heart.csv']


We start by reading the data from the CSV File.
We then split the data into inputs and targets.

In [6]:
df = pd.read_csv("../input/heart-disease-uci/heart.csv")
np_arr = df.values
X = np_arr[:, 0:13]
Y = np_arr[:, 13]
print(X.shape, Y.shape)

(303, 13) (303,)


We are using StandardScaler from sklearn library to process the input data.

In [7]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(X)
s_X = scaler.transform(X)

We make the Sequential model using Keras.layers
The Network is dense connected and of shape [13, 10, 5, 1]

We use Adam optimizer, mean_squared_loss and accuracy as evaluation metric.

In [8]:
model = keras.Sequential()
model.add(layers.Dense(10, activation = 'relu', input_shape = (13,)))
model.add(layers.Dense(5, activation = 'relu'))
model.add(layers.Dense(1, activation = 'sigmoid'))

model.compile(optimizer = 'Adam', loss = 'mean_squared_error', metrics=['accuracy'])

We fit the dataset into the model.

Using 500 epochs and batches of size 10

In [ ]:
history = model.fit(s_X, Y, epochs = 500, batch_size = 10, verbose=0)

Epoch 1/500
303/303 [==============================] - 1s 2ms/step - loss: 0.2917 - acc: 0.4719
Epoch 2/500
303/303 [==============================] - 0s 177us/step - loss: 0.2680 - acc: 0.5182
Epoch 3/500
303/303 [==============================] - 0s 168us/step - loss: 0.2527 - acc: 0.5578
Epoch 4/500
303/303 [==============================] - 0s 160us/step - loss: 0.2403 - acc: 0.6370
Epoch 5/500
303/303 [==============================] - 0s 166us/step - loss: 0.2279 - acc: 0.6832
Epoch 6/500
303/303 [==============================] - 0s 166us/step - loss: 0.2153 - acc: 0.7228
Epoch 7/500
303/303 [==============================] - 0s 151us/step - loss: 0.2018 - acc: 0.7624
Epoch 8/500
303/303 [==============================] - 0s 155us/step - loss: 0.1884 - acc: 0.7855
Epoch 9/500
303/303 [==============================] - 0s 162us/step - loss: 0.1762 - acc: 0.7921
Epoch 10/500
303/303 [==============================] - 0s 167us/step - loss: 0.1648 - acc: 0.8020
Epoch 11/500
303/303 

303/303 [==============================] - 0s 161us/step - loss: 0.0720 - acc: 0.9208
Epoch 84/500
303/303 [==============================] - 0s 158us/step - loss: 0.0718 - acc: 0.9208
Epoch 85/500
303/303 [==============================] - 0s 156us/step - loss: 0.0710 - acc: 0.9208
Epoch 86/500
303/303 [==============================] - 0s 153us/step - loss: 0.0708 - acc: 0.9208
Epoch 87/500
303/303 [==============================] - 0s 171us/step - loss: 0.0703 - acc: 0.9208
Epoch 88/500
303/303 [==============================] - 0s 158us/step - loss: 0.0699 - acc: 0.9208
Epoch 89/500
303/303 [==============================] - 0s 158us/step - loss: 0.0697 - acc: 0.9208
Epoch 90/500
303/303 [==============================] - 0s 153us/step - loss: 0.0692 - acc: 0.9241
Epoch 91/500
303/303 [==============================] - 0s 169us/step - loss: 0.0690 - acc: 0.9241
Epoch 92/500
303/303 [==============================] - 0s 153us/step - loss: 0.0686 - acc: 0.9274
Epoch 93/500
303/303 [=

303/303 [==============================] - 0s 157us/step - loss: 0.0468 - acc: 0.9571
Epoch 166/500
303/303 [==============================] - 0s 155us/step - loss: 0.0469 - acc: 0.9571
Epoch 167/500
303/303 [==============================] - 0s 150us/step - loss: 0.0465 - acc: 0.9538
Epoch 168/500
303/303 [==============================] - 0s 159us/step - loss: 0.0462 - acc: 0.9571
Epoch 169/500
303/303 [==============================] - 0s 154us/step - loss: 0.0458 - acc: 0.9571
Epoch 170/500
303/303 [==============================] - 0s 152us/step - loss: 0.0458 - acc: 0.9538
Epoch 171/500
303/303 [==============================] - 0s 162us/step - loss: 0.0454 - acc: 0.9538
Epoch 172/500
303/303 [==============================] - 0s 152us/step - loss: 0.0452 - acc: 0.9538
Epoch 173/500
303/303 [==============================] - 0s 168us/step - loss: 0.0450 - acc: 0.9571
Epoch 174/500
303/303 [==============================] - 0s 164us/step - loss: 0.0449 - acc: 0.9571
Epoch 175/500


303/303 [==============================] - 0s 157us/step - loss: 0.0335 - acc: 0.9703
Epoch 248/500
303/303 [==============================] - 0s 146us/step - loss: 0.0333 - acc: 0.9703
Epoch 249/500
303/303 [==============================] - 0s 153us/step - loss: 0.0331 - acc: 0.9703
Epoch 250/500
303/303 [==============================] - 0s 151us/step - loss: 0.0332 - acc: 0.9703
Epoch 251/500
303/303 [==============================] - 0s 158us/step - loss: 0.0327 - acc: 0.9703
Epoch 252/500
303/303 [==============================] - 0s 150us/step - loss: 0.0328 - acc: 0.9703
Epoch 253/500
303/303 [==============================] - 0s 160us/step - loss: 0.0326 - acc: 0.9703
Epoch 254/500
303/303 [==============================] - 0s 167us/step - loss: 0.0326 - acc: 0.9703
Epoch 255/500
303/303 [==============================] - 0s 163us/step - loss: 0.0327 - acc: 0.9703
Epoch 256/500
303/303 [==============================] - 0s 156us/step - loss: 0.0324 - acc: 0.9703
Epoch 257/500


303/303 [==============================] - 0s 148us/step - loss: 0.0265 - acc: 0.9769
Epoch 330/500
303/303 [==============================] - 0s 150us/step - loss: 0.0265 - acc: 0.9769
Epoch 331/500
303/303 [==============================] - 0s 156us/step - loss: 0.0264 - acc: 0.9769
Epoch 332/500
303/303 [==============================] - 0s 156us/step - loss: 0.0264 - acc: 0.9769
Epoch 333/500
303/303 [==============================] - 0s 152us/step - loss: 0.0263 - acc: 0.9769
Epoch 334/500
303/303 [==============================] - 0s 157us/step - loss: 0.0263 - acc: 0.9769
Epoch 335/500
303/303 [==============================] - 0s 159us/step - loss: 0.0265 - acc: 0.9769
Epoch 336/500
303/303 [==============================] - 0s 177us/step - loss: 0.0263 - acc: 0.9769
Epoch 337/500
303/303 [==============================] - 0s 167us/step - loss: 0.0262 - acc: 0.9769
Epoch 338/500
303/303 [==============================] - 0s 155us/step - loss: 0.0261 - acc: 0.9769
Epoch 339/500


Finally the model has been trained. Now we can visualize the training and view accuracy and confusion matrix.

In [ ]:
y_pred = model.predict(s_X)
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
cm = confusion_matrix(Y, y_pred.round())
acc = accuracy_score(Y, y_pred.round())
print("Accuracy of the model : ",acc)
print("The Confusion matrix :\n", cm)

Now Finally we can take a look at the loss value throughout the training.

In [ ]:
import matplotlib.pyplot as plt

plt.plot(history.history['loss'])
plt.title('Model loss')
plt.ylabel('Loss')
plt.xlabel('Epoch')
plt.legend(['Train'], loc='upper right')
plt.show()

We now have a model that can predict the Heart Diseases with about 97% accuracy.